### PR TITLE
Common properties for NuGet packages

### LINKED ISSUE #6404: Description for NuGet packages
All packages just have "Package Description" now. However, we don't want to write descriptions for all individual packages (unless someone is masochistic enough) so a better thing would be to generate it somehow.

- For projects with a Manifest include the descriptions from there.
- For other projects use a common template.

In both cases, the text could include something like:

> Orchard Core is an open-source, modular, multi-tenant application framework and CMS for ASP.NET Core. For detailed information about Orchard Core, and for documentation on how to use it and integrate it into your app see https://www.orchardcore.net/.

### PR DESCRIPTION
**Pull request by ns8482e**
### Todo ( In separate PR )
- All Module title & description.
- Publish Symbol package to nuget.org and use of nuget v3 api in appveyor.yml for master branch (or for dev branch if cloudsmith.io supports symbol package)

### Completed
- Added Common Package description
- Update build script in appveyor.yml  and travis.yml for passing properties `SourceRevisionId`,  `GitBranch` to `dotnet build` to support Source Link
- Themes
   - [x] TheAdmin, 
   - [x] TheAgency
   - [x]  TheCommingSoon
   - [x] TheBlog 

- Add Source Link Properties 
- Added Git Commit hash for Product Version and build by passing git-commit-hash as in following command

Script to include git commit hash during build
```bash
dotnet build -c Release /p:SourceRevisionId=c990b5c38f0a44fb4f2ecb53b9848f8f4ceba0d0  /p:GitBranch=dev
dotnet pack -c Release --no-build
```
### Generated DLL Properties
image

### Generated Package Properties
image

Fixes #6404

### CODE REVIEW

**Comment by agriffard** [comment]
Please move the remaining Todo tasks at the top with empty checkboxes or tell us if it will be in other PRs.

Do we already see the new Nuget properties on the Preview package source?

**Comment by ns8482e** [comment]
> Please move the remaining Todo tasks at the top with empty checkboxes or tell us if it will be in other PRs.
>
Done 


> Do we already see the new Nuget properties on the Preview package source?
>

### Package metadata - Before (TheAdmin)
image


### Package metadata - After this PR (TheAdmin)
image

File: .travis.yml [hunk 1]
```diff
@@ -30,7 +30,7 @@ script:
   - if test "$TRAVIS_OS_NAME" == "linux"; then export PATH="/home/travis/.dotnet":"$PATH"; fi
   - if test "$TRAVIS_OS_NAME" == "osx"; then export PATH="/Users/travis/.dotnet":"$PATH"; fi
   - dotnet --info
-  - dotnet build -c Release
+  - dotnet build -c Release /p:SourceRevisionId=$TRAVIS_COMMIT /p:GitBranch=$TRAVIS_BRANCH
```
Comment (sebastienros) [inline .travis.yml:33]: Why is that necessary?

File: appveyor.yml [hunk 2]
```diff
@@ -50,7 +50,7 @@ install:
 # - ps: .\dotnet-install.ps1 --Version 3.1.402
 
 build_script:
-  - dotnet build -c Release
+  - dotnet build -c Release /p:SourceRevisionId=$env:APPVEYOR_REPO_COMMIT /p:GitBranch=$env:APPVEYOR_REPO_BRANCH
```
Comment (sebastienros) [inline appveyor.yml:53]: Isn't the plugin able to find these by itself?

**Comment by sebastienros** [comment]
Can we add a description for OrchardCore.Localization.Core and .Abstractions? These are packages that are referenced directly too.

**Comment by sebastienros** [comment]
Cloudsmith doesn't support v3

File: .travis.yml (hunk 1 above)
Comment (ns8482e) [reply .travis.yml:33]: To pass `props` to `dotnet build` to use current commit and current branch of executing CI that available as CI variables.
 These  `props` are defined in `OrchardCore.Commons.props`  to generate `commit` attribute for nuget package.

`SourceRevisionId` prop is used to have `Product Version` in DLL properties.

File: appveyor.yml (hunk 2 above)
Comment (ns8482e) [reply appveyor.yml:53]: Same as above. current commit and current branch is available as variable in CI

**Comment by lskillen** [comment]
> Cloudsmith doesn't support v3

Is this the NuGet API, or something else? It definitely supports the v3 API; but if something else is missing, or not quite right, we'd be happy to look at it for you. :-)

**Comment by ns8482e** [comment]
> Can we add a description for OrchardCore.Localization.Core and .Abstractions? These are packages that are referenced directly too.

In this PR, common NuGet package properties are defined - i.e. `OrchardCore.Commons.props` which is applicable to all projects via `Directory.Build.props`. 

Title and Description is unique to each project that I'll cover in separate PR.

File: src/OrchardCore.Build/OrchardCore.Commons.props [hunk 3]
```diff
@@ -16,11 +16,29 @@
...
+    <SymbolPackageFormat>snupkg</SymbolPackageFormat>
     
     <!-- This is used by GitHub Repository to find which repository should contain the package when updloaded -->
     <RepositoryUrl>https://github.com/OrchardCMS/OrchardCore</RepositoryUrl>
+    <PackageProjectUrl>https://www.orchardcore.net/</PackageProjectUrl>
+    <RepositoryCommit>$(GitCommit)</RepositoryCommit>
+    <RepositoryBranch>$(GitBranch)</RepositoryBranch>
     
+    <PackageTags>OrchardCore OrchardCoreCMS CMS MVC AspNetCore</PackageTags>
+    <Copyright>Orchard Core CMS Community and Contributors</Copyright>
```
Comment (lahma) [inline src/OrchardCore.Build/OrchardCore.Commons.props:36]: Should there also be a license expression here (BSD 3-Clause) if it's shared will all the packages, at least repo root declares LICENSE with BSD 3-Clause

Comment (ns8482e) [reply src/OrchardCore.Build/OrchardCore.Commons.props:36]: Probably embed the actual license file from repo?

Comment (lahma) [reply src/OrchardCore.Build/OrchardCore.Commons.props:36]: NuGet.org supports expressions to show correct link in the UI

Comment (ns8482e) [reply src/OrchardCore.Build/OrchardCore.Commons.props:37]: Done

**Comment by ns8482e** [comment]
> 
> Is this the NuGet API, or something else? It definitely supports the v3 API; but if something else is missing, or not quite right, we'd be happy to look at it for you. :-)

For publishing `snupkg`, To support source link.

https://docs.microsoft.com/en-us/nuget/create-packages/symbol-packages-snupkg#publishing-a-symbol-package  As per this link for NuGet.org - requires V3 API - for (master branch)

Idea is if  Cloudsmith.io io supports  `snupkg` V3 API, we could have source link for dev branch too.

**Comment by jtkech** [comment]
Do we really need source link?

As i remember our packages include their pdb files that embed the related source files, see #1521

**Comment by ns8482e** [comment]
> Do we really need source link?
> 
> As i remember our packages include their pdb files that embed the related source files, see #1521

@jtkech  #1521 is Nice to know we already have debug support  👍 

But IMHO when source link is released an year ago and NuGet already supports `snupkg` It  better not to include `pdb` and source code embedded in same `NuGet` package. 

Also For Release 1.0 and production ready version of orchard - its better if  we keep symbols separate from main package in `snupkg`, and with source link we don't need to embed the source code in package - reducing size of packages.

**Comment by jtkech** [comment]
@ns8482e okay, i'm not against

But for info the package size doesn't grow too much because the pdb are compressed before being embedded in the dll

**Comment by ns8482e** [comment]
> @ns8482e okay, i'm not against
> 
> But for info the package size doesn't grow too much because the pdb are compressed before being embedded in the dll

Ok, when I proposed `snupkg` here,  I wasn't aware that source debugging is already supported. I always learn something new from you. 

I am referring following link for source link related info
https://github.com/dotnet/sourcelink . 

According to document,
> 
> Prior availability of NuGet.org symbol server the recommendation used to be to include the PDB in the main NuGet package by setting the following property in your project:
> 
> ```
>  <PropertyGroup>
>      <AllowedOutputExtensionsInPackageBuildOutputFolder>$(AllowedOutputExtensionsInPackageBuildOutputFolder);.pdb</AllowedOutputExtensionsInPackageBuildOutputFolder>
> </PropertyGroup>  
> ```
> Including PDBs in the .nupkg is generally no longer recommended ....
>.....
> Consider including PDBs in the main package only if it is not possible to use .snupkg for the above reasons. For managed projects, consider switching to Portable PDBs by setting DebugType property to portable.

So, If embedding source/pdb is not longer recommended, then It would be nice to opt for github source link and `snupkg` towards release 1.0 of orchard core.

**Comment by jtkech** [comment]
@ns8482e no problem i'm open to any solution, it's just for infos

Hmm, but yes i like our current standalone solution ;)

Yes, in the readme they said that it is not recommended but precisely for the size, also they talk about Visual Studio 2017, maybe in the meantime it has been improved, as said it seems that data are compressed.

So maybe we could compare the package sizes and see if it is worth to change our current debug type, knowing that we also embed asset files, and that we also have `*.Views.dll` for precompiled views.

Anyway, you will have to take into account the `Directory.Build.props / targets` files that we have under the parent folders of core projects / modules and themes, and where we already set e.g. the `DebugType`  property, and where we already use the `AllowedOutputExtensionsInPackageBuildOutputFolder` property.

**Comment by ns8482e** [comment]
@jtkech removed `AllowedOutputExtensionsInPackageBuildOutputFolder` as you suggested.
Here is the size comparison.

| Packages   | Embedded Source   | Not Embeded|
| ----------- | ----------- | ----------- |
| 143 Packages <br/>(Excluding following 4 packages ) | 11.6 MB (12,259,700 bytes)| <b>6.90 MB </b>(7,243,689 bytes)| 
|TheComingSoonTheme|9.49 MB (9,959,487 bytes)| 9.48 MB (9,945,842 bytes)|
|TheBlogTheme|8.29 MB (8,693,711 bytes)|8.27 MB (8,679,230 bytes)|
|OrchardCore.Resources|5.27 MB (5,536,155 bytes)| 5.25 MB (5,513,832 bytes)|
|TheAgencyTheme|3.56 MB (3,742,826 bytes)|3.55 MB (3,728,093 bytes)|

File: .travis.yml (hunk 1 above)
Comment (ns8482e) [reply .travis.yml]: @sebastienros Using `IncludeSourceRevisionInInformationalVersion`

**Comment by jtkech** [comment]
@ns8482e cool!!!

Hmm, 11.6 - 6.9 = 4.7, so we have a difference of 4.7Mb but for 143 packages (not per package) right?

If so, is it still worth to not keep embedded sources?

Need to leave, , i will look at it tomorrow

**Comment by ns8482e** [comment]
> Hmm, 11.6 - 6.9 = 4.7, so we have a difference of 4.7Mb but for 143 packages (not per package) right?
> 

Yes it’s total size for 143 packages.

**Comment by jtkech** [comment]
@ns8482e 

Okay, i know there are long and still open issues on the dotnet repo around this, would need more time to see again all pros and cons, so i'm open to both solutions.

If no other feedback, do it as you think it is better ;)

**Comment by agriffard** [comment]
So @jtkech, can we merge this PR?

**Comment by jptissot** [comment]
@agriffard I would wait. We will loose symbols with cloudsmith if we merge this now as they do not yet support symbol server.

**Comment by jtkech** [comment]
@agriffard 

I let you review all the texts, e.g.

    <Copyright>Orchard Core CMS Community and Contributors</Copyright>
    <Authors>Orchard Core CMS Community and Contributors</Authors>
    <Owners>Orchard Core CMS Community and Contributors</Owners>

E.g. do we need the term `CMS` as OrchardCore is also a framework, so maybe

    <Copyright>OrchardCore Community and Contributors</Copyright>

Then i let others decide on publishing symbol packages in place of embedding them, looks good, i'm not against

But as @jptissot said, symbol server is not yet supported by cloudsmith

File: src/OrchardCore.Build/OrchardCore.Commons.props [hunk 4]
```diff
@@ -17,10 +17,27 @@
     <LangVersion>8.0</LangVersion>
 
     <!-- Common NuGet properties-->
+    <EmbedUntrackedSources>true</EmbedUntrackedSources>
+    <IncludeSourceRevisionInInformationalVersion>true</IncludeSourceRevisionInInformationalVersion>
+    <ContinuousIntegrationBuild Condition="'$(CI)' == 'true'">true</ContinuousIntegrationBuild>
```
Comment (sebastienros) [inline src/OrchardCore.Build/OrchardCore.Commons.props:22]: Explain?

File: src/OrchardCore.Build/OrchardCore.Commons.props [hunk 5]
```diff
@@ -17,10 +17,27 @@
...
+
+    <PublishRepositoryUrl>true</PublishRepositoryUrl>
```
Comment (sebastienros) [inline src/OrchardCore.Build/OrchardCore.Commons.props:24]: Explain?

File: src/OrchardCore.Build/OrchardCore.Commons.props [hunk 6]
```diff
@@ -17,10 +17,27 @@
...
+    <!--This is used to generate symbol package for Source Link-->
+    <IncludeSymbols>true</IncludeSymbols>
+    <SymbolPackageFormat>snupkg</SymbolPackageFormat>
     
     <!-- This is used by GitHub Repository to find which repository should contain the package when updloaded -->
     <RepositoryUrl>https://github.com/OrchardCMS/OrchardCore</RepositoryUrl>
+    <PackageProjectUrl>https://www.orchardcore.net/</PackageProjectUrl>
+    <PackageLicenseExpression>BSD-3-Clause</PackageLicenseExpression>
     
+    <PackageTags>OrchardCore MVC AspNetCore</PackageTags>
```
Comment (sebastienros) [inline src/OrchardCore.Build/OrchardCore.Commons.props:34]:
```suggestion
    <PackageTags>OrchardCore</PackageTags>
```

File: src/OrchardCore.Themes/TheAdmin/TheAdmin.csproj [hunk 7]
```diff
@@ -3,6 +3,11 @@
   <PropertyGroup>
     <TargetFramework>$(AspNetCoreTargetFramework)</TargetFramework>
     <AddRazorSupportForMvc>true</AddRazorSupportForMvc>
+
+    <!-- NuGet properties-->
+    <Title>The Admin theme</Title>
+    <PackageTags>$(PackageTags) TheAdmin Theme</PackageTags>
```
Comment (sebastienros) [inline src/OrchardCore.Themes/TheAdmin/TheAdmin.csproj:9]:
```suggestion
    <PackageTags>$(PackageTags) Theme</PackageTags>
```

File: src/OrchardCore.Themes/TheAgencyTheme/TheAgencyTheme.csproj [hunk 8]
```diff
@@ -3,6 +3,11 @@
   <PropertyGroup>
     <TargetFramework>$(AspNetCoreTargetFramework)</TargetFramework>
     <AddRazorSupportForMvc>true</AddRazorSupportForMvc>
+
+    <!-- NuGet properties-->
+    <Title>The Agency Theme</Title>
+    <PackageTags>$(PackageTags) TheAgency Theme</PackageTags>
```
Comment (sebastienros) [inline src/OrchardCore.Themes/TheAgencyTheme/TheAgencyTheme.csproj:9]:
```suggestion
    <PackageTags>$(PackageTags) Theme</PackageTags>
```

File: src/OrchardCore.Themes/TheBlogTheme/TheBlogTheme.csproj [hunk 9]
```diff
@@ -3,6 +3,11 @@
   <PropertyGroup>
     <TargetFramework>$(AspNetCoreTargetFramework)</TargetFramework>
     <AddRazorSupportForMvc>true</AddRazorSupportForMvc>
+
+    <!-- NuGet properties-->
+    <Title>The Blog Theme</Title>
+    <PackageTags>$(PackageTags) TheBlog Theme</PackageTags>
```
Comment (sebastienros) [inline src/OrchardCore.Themes/TheBlogTheme/TheBlogTheme.csproj:9]:
```suggestion
    <PackageTags>$(PackageTags) Theme</PackageTags>
```

File: src/OrchardCore.Themes/TheComingSoonTheme/TheComingSoonTheme.csproj [hunk 10]
```diff
@@ -3,6 +3,11 @@
   <PropertyGroup>
     <TargetFramework>$(AspNetCoreTargetFramework)</TargetFramework>
     <AddRazorSupportForMvc>true</AddRazorSupportForMvc>
+    
+    <!-- NuGet properties-->
+    <Title>The Coming Soon Theme</Title>
+    <PackageTags>$(PackageTags) TheComingSoon Theme</PackageTags>
```
Comment (sebastienros) [inline src/OrchardCore.Themes/TheComingSoonTheme/TheComingSoonTheme.csproj:9]:
```suggestion
    <PackageTags>$(PackageTags) Theme</PackageTags>
```

**Comment by sebastienros** [comment]
@lskillen sorry, didn't notice your question.

When we moved to CS, I tried the v3 api endpoint, but it didn't work as a "push" packages argument. Pretty sure I used the correct url since I have done that many times on NuGet and MyGet for other projects. AFAIR the consumption operations worked fine. It only started working when I used the v2 api endpoint.

**Comment by sebastienros** [comment]
Let's drop the CMS work in the descriptions for now. Only if defined in specific modules and meta packages that are specific for CMS, maybe later.

**Comment by lskillen** [comment]
@sebastienros That's odd! Happy to help debug to see if something we need to fix? I see threads about the symbol server as well. We didn't implement it because there was a lack of demand from those we asked at the time of implementation, but certainly happy to re-think prioritisation on that too. :-) You can email me anytime (lee @ our site domain).

File: src/OrchardCore.Build/OrchardCore.Commons.props (hunk 5 above)
Comment (ns8482e) [reply src/OrchardCore.Build/OrchardCore.Commons.props:24]: setting as per doc [here](https://github.com/dotnet/sourcelink/blob/master/docs/README.md#publishrepositoryurl)

File: src/OrchardCore.Build/OrchardCore.Commons.props (hunk 4 above)
Comment (ns8482e) [reply src/OrchardCore.Build/OrchardCore.Commons.props:22]: setting as per doc [here](https://github.com/dotnet/sourcelink/blob/master/docs/README.md#continuousintegrationbuild)

**Review by sebastienros — APPROVED** [review]
